### ISSUE TITLE
jspm install times out and 100% CPU usage

### ISSUE THREAD

**Issue by martingust** [jspm/jspm-cli#675]
I’m having trouble `jspm install` on our CI server (Windows Server 2012). 9/10 times I get `Timed out on lookup for github:aurelia/…`. 

When looking at the task manager, new `Git for Windows` processes keeps piling up, and the CPU is running on 100% until most of the packages have timed out.

- I have tried to increase the timeout which does not help.
- GitHub credentials are configured.
- The server is behind a corporate firewall, but I can’t see anything get stuck (I could be wrong), but a few of the packages usually get downloaded.
- The network connection is good, the server's hardware should not be the problem either.
- Featching same packages with Bower over HTTPS works fine.

I realise the problem most likely is on my end, but have no clue really how to continue troubleshooting this, any ideas?

**Comment by guybedford** [comment]
If you're getting a timeout, this may be due to instability on GitHub or due to running over an internal proxy.

Is it still giving issues? Are you sure you don't need a HTTPS proxy at all? Are your credentials definitely showing up as testing ok?

**Comment by guybedford** [comment]
@martingust did you make any progress here?

**Comment by martingust** [comment]
@guybedford Credentials are ok and no need for proxy. We increased timeout further and added a second core to the processor, that gets the packages downloaded but still obnoxiously slow and CPU running on 100% for long time. Might need even more CPU, also going to try with node 0.10.x oppose to 0.12.x which seemed to solve similar problems for others.

**Comment by guybedford** [comment]
@martingust thanks for the update. Perhaps if you could share the exact install and I'll check it myself in Windows. Is it just the Aurelia skeleton?

**Comment by deveddie** [comment]
Hi, I'm also experiencing the same issue with the same repository Seeing 99% CPU, a jump from 20 - 77% memory and lots of "Git for Windows" processes in the task manager :(

**Comment by guybedford** [comment]
This is just common operation - jspm performs parallel downloading over the tree.

There are build operations that are applied when installing a package from npm, and sometimes jspm can end up doing more processing than it needs here.

If you're specifically installing a package from npm that is taking a while I can look into it to try and assist with the right override to reduce the compilation steps, but otherwise this is just normal operation.

**Comment by deveddie** [comment]
Chucking more CPU and memory at the server and increasing the timeout on jspm seems to have fixed the issue - it does seem though that jspm isn't giving the server time to breath though and so without enough memory / processing power to complete each task in a timely manner it implodes on itself - kind of like a tight loop without time for garbage collection/event handling.

**Comment by martingust** [comment]
@guybedford This is a custom Aurelia app, but we did have the same problems when testing with Aurelia skeleton.

@deveddie Did this fix all problems, do they download fast and not 100% CPU? Or did it just fixed the timeout issue?

**Comment by guybedford** [comment]
One option here might be to start pooling the Git callouts potentially, I created an issue for this in https://github.com/jspm/github/issues/42.

Note that implementing https://github.com/jspm/github/issues/25 directly would potentially alleviate this as well as the other issues with calling out to git.

**Comment by deveddie** [comment]
@martingust The server was still very stressed and I received a good number of timeout warnings but it did eventually complete in one go. The same install caused almost no issues on my laptop pc, only on the vm server I use for dev work. Luckily as it was a vm I could supply it with more resources easily.

**Comment by guybedford** [comment]
@deveddie it could be worth ensuring you've updated to the most recent version of git in the path as well here as that may well be where these performance gains are to be found. We're basically running dozens of `git ls-remote` operations side-by-side.

**Comment by deveddie** [comment]
@guybedford Yip, i'd agree but both my local machine and the VM server had the latest version of git loaded on them yesterday morning direct from the git website.

**Comment by amcdnl** [comment]
@guybedford any updates on this?  I'm encountering this issue as well.  I'm running Windows Server 2012 on my builder server too.

**Comment by lookfirst** [comment]
@amcdnl Don't use windows! :stuck_out_tongue_winking_eye:

**Comment by deveddie** [comment]
Some of us are forced to by our evil overlords :(

**Comment by amcdnl** [comment]
Its my server, I don't use locally ;)

**Comment by guybedford** [comment]
@amcdnl I can't fix this until I'm able to replicate it unfortunately. Alternatively I can run proposed fixes and have you test them against your own server, but it will be a long process to do that type of feedback loop.

Just to go over the basics (and apologies if this is repeating):
* Can you confirm you're running the latest version of Git?
* Can you let me know what version of Windows this is running? Is it just a execution environment?

It would help to know so I can try and replicate in a VM. Thanks.

**Comment by amcdnl** [comment]
Can you confirm you're running the latest version of Git?
Yes.  First thing I checked when reading this forum.

Can you let me know what version of Windows this is running? Is it just a execution environment?
This is the build server that compiles the code and deploys ( Jenkins ).  Windows 2012 Server.

@martingust @deveddie are you guys still getting this too?

**Comment by guybedford** [comment]
@amcdnl can you confirm that a normal jspm install on a Windows 2012 server is all that is needed to replicate? Or do I need to specifically be running through a special environment?

**Comment by martingust** [comment]
@amcdnl I still have the issue, but added a third core to the VM and that gets through the install reasonably fast. 

@guybedford That would not be all needed. I have tried jspm install on a dedicated Windows 2012 Server, 6 core processor, on a different network and it behaves normal.

**Comment by deveddie** [comment]
Yep, sorry, have been fighting bugs. I was going to say I think the common thread here is that all the problems seem to be on VMs.

**Comment by guybedford** [comment]
Thanks - I can't do anything to assist here until someone can explain how I can replicate it.

Will leave this up for discussion in the mean time.

**Comment by amcdnl** [comment]
@guybedford Probably a windows server 2012 VM could do it.  you can download a trial version.

@martingust I bumped my VM to 8 cores and had no luck. :(

**Comment by ejsmith** [comment]
This is driving me insane. I think it would be pretty easy to setup a public repo that does a jspm install on appveyor. Would that help show the problem and figure out what is going on?

**Comment by sujesharukil** [comment]
Having the same problem here. This is really killing me. I have one week to evaluate and move on. My build server is an Azure VM running Windows Server 2012, 4 core, 7 GB. jspm install times out. I am trying to get aurelia-skeleton working. After spending a week working with the hosted build controller on VSO, I moved it to an on premise. Works well on my dev machine, but the VM, not so. You can easily reproduce this by just trying to do a jspm install -y on aurelia-skeleton on an Azure vm.

**Comment by martingust** [comment]
@guybedford According to everybody the way to replicate would be: 

`jspm install` Aurelia Skeleton Navigation on a VM with Windows Server 2012.

**Comment by sujesharukil** [comment]
@guybedford do you need any more information on this? This issue is the last standing issue for me. I will have to move on to something else to get Aurelia loading. I really want to use jspm.

**Comment by crisptrutski** [comment]
@martingust We've [made a branch](https://github.com/jspm/github/tree/git-pool) to test if execution pooling resolves your issue.

Would you mind testing out on your side? Haven't reproduced the issue on my side yet.

Assuming a local jspm install, swapping in this branch could be as simple as:

```
rm -rf node_modules/jspm/node_modules/jspm-github
git clone -b git-pool git@github.com:jspm/github.git node_modules/jspm/node_modules/jspm-github
```

Otherwise I'll set up a VM tomorrow morning and hopefully reproduce the issue

**Comment by martingust** [comment]
@crisptrutski That is exiting. But after swapping in the branch I'm getting this error on `jspm install`:
```
C:\GitHub\skeleton-navigation>jspm install -y

err  Error: Cannot find module 'tar'
         at Function.Module._resolveFilename (module.js:338:15)
         at Function.Module._load (module.js:280:25)
         at Module.require (module.js:364:17)
         at require (module.js:380:17)
         at Object.<anonymous> (C:\GitHub\skeleton-navigation\node_modules\jspm\
node_modules\jspm-github\github.js:11:11)
         at Module._compile (module.js:456:26)
         at Object.Module._extensions..js (module.js:474:10)
         at Module.load (module.js:356:32)
         at Function.Module._load (module.js:312:12)
         at Module.require (module.js:364:17)

err  Unable to load registry github

warn Installation changes not saved.
```

**Comment by crisptrutski** [comment]
@martingust thanks for jumping in - realise there was a step missing from the instructions I gave: running `npm install` inside the `jspm-github` checkout.

The missing transitive deps are causing failure to load the registry

**Comment by xFrigginx** [comment]
I had the same problem on a windows 8.1 machine. After disabling Windows Endpoint Security 10, everthing went down smoothly.

**Comment by crisptrutski** [comment]
@guybedford can't think of a solution here. 

Perhaps outputting a "Antivirus may hurt you, blah blah" warning during the git execution if it's taking unusually long, Loath to fail the action or sniff for antivirus services directly.

Are we still sure this is git related? Could be unzipping too.

Also surprised that CI servers are running antivirus software.. Perhaps we've still got a separate issue at play here

**Comment by crisptrutski** [comment]
@martingust anything further here? did you test again after install the registry dependencies? can you confirm or deny whether there is an antivirus on the machine?

**Comment by martingust** [comment]
@crisptrutski I just got to try again, unfortunately no luck with execution pooling, and we do not have any antivirus on the machine.

**Comment by martingust** [comment]
@crisptrutski I just got to try this again. No luck after installing the registry dependencies, still times out. However we do have `System Center Endpoint Protection` on the server. I'm going to look into how that can be disabled and let you know.

**Comment by crisptrutski** [comment]
@martingust thanks for testing out - good to know that pooling is not the solution

**Comment by martingust** [comment]
@crisptrutski I disabled antivirus no differnce.

**Comment by pawelpawlow** [comment]
When installing aurelia packages on Windows, jspm starts multiple "Git for Windows (32 bit)" processes  (more than 20). This freezes the machine and eventually may lead to lookup timeouts. It looks like the real problem that jspm starts too many simultaneous instances of git for windows.
What can I do to limit the number of git instances?
I'm running the latest releases of jspm and git for windows on Windows Server 2012 R2 64-bit.

**Comment by guybedford** [comment]
@pawelpawlow we implemented execution pooling in https://github.com/jspm/jspm-cli/issues/675#issuecomment-103919249 to see if that helps the issue. If you're able to test this and let us know that would be really helpful. The hardest part of this issue is very much replicating the problem to fix it.

**Comment by pawelpawlow** [comment]
@guybedford ok, will give it a try - thank you

**Comment by pawelpawlow** [comment]
@guybedford git-pool branch still freezes the machine and ends with timeouts; I still see a lot of git processes running!? How many parallel gits should this branch of jspm run?

**Comment by guybedford** [comment]
The pool is supposed to limit to 5 requests. If there are still a lot running that may be a problem with the pooling. Unfortunately @crisptrutski is on client work at the moment and isn't able to look at this again right now. I can aim to look into the pooling again as soon as this release is out, but am currently focused on this release. Sorry can't be of more help right now.

**Comment by pawelpawlow** [comment]
Those git processes keep running after jspm finishes! I think there some 40-50 of them still running.

**Comment by pawelpawlow** [comment]
Ok, I'll wait for @crisptrutski then :)

**Comment by guybedford** [comment]
Interesting, I never see more than 10 processes on my Mac, and they all shut down fine. Will definitely investigate this further soon, it's just finding the time at this point. Thanks for patience.

**Comment by pawelpawlow** [comment]
@crisptrutski Looks like pool's promise doesn't wait for the child_process.exec's callback to be called and assumes that as soon as child_process.exec() returns it's ok to fire the next one.

**Comment by XVincentX** [comment]
I am having the same problem when restoring packages on heroku dynos and azure websites as well.

**Comment by vanko-dev** [comment]
similarly on teamcity

**Comment by donald-slagle** [comment]
I'm having the same issue on a Rackspace server with 4 cores and 2gb of ram. We are running Windows Server 2008 R2  and have no problems building asp.net apps with many nuget packages. Running `jspm install` our server CPU maxes out and the server becomes completely unusable until all the packages timeout.

**Comment by brendonparker** [comment]
I'm having the same problem. It's an aurelia app. Fires up a bunch of git processes, pegs the CPU at 100%, eventually they start timing out. 

Running an AWS instance Server 2012 R2 (t2.small). Single Core, 2.5 GHz, 2 GB RAM.

**Comment by curran** [comment]
It seems like this may be caused by JSPM launching all the install processes at once, rather than setting a limit on the number of concurrent processes. I think the culprit may be at [install.js line 116](https://github.com/jspm/jspm-cli/blob/master/lib/install.js#L116), where Promise.all is being used to fire off all the install processes (at least I think that's what this line is doing, not 100% sure).

This project [es6-promise-pool](https://www.npmjs.com/package/es6-promise-pool) implements a variant of Promise.all where you can limit the maximum number of simultaneous operations. Using this or something similar as a drop-in replacement for Promise.all for launching installer tasks may solve this issue.

**Comment by guybedford** [comment]
There was a pooling test in https://github.com/jspm/github/tree/git-pool, which should have caught this, but it would be good to look into that work again.

**Comment by brendonparker** [comment]
The pooling branch worked for me, after I modified the code to do this:
````javascript
function gitExec(command, execOpt, callback) {
  return gitPool.execute(function() {
    return new Promise(function(resolve){
        exec('git ' + command, execOpt, function(err, stdout, stderr){
            callback(err, stdout, stderr);
            resolve();
        });
    });
  });
}
````
instead of:
````javascript
function gitExec(command, execOpt, callback) {
  return gitPool.execute(function() {
    exec('git ' + command, execOpt, callback);
  });
}
````

I also had to dial back the pool size to 2.

I can add a PR for that branch, if someone wants to take a look.

**Comment by guybedford** [comment]
@brendonparker that is amazing to hear! Thanks for trying this out. A PR with the fix would be incredibly welcome. Also it would be useful to have an environment detection to detect the pool size, as a larger pool size is preferable on unix machines. Just windows? Any better suggestion here is welcome.

**Comment by brendonparker** [comment]
You can see the comments in the pull request - I'm not sure which branch you'd want to consider it for (the git-pool or master). I believe this is only a problem on windows platforms, so I just ditched the pool altogether for non-windows platforms.

**Comment by guybedford** [comment]
A massive thanks to @crisptrutski for the pooling work and @brendonparker for testing this out and getting it in. The pooling solution has been released on the jspm beta (`npm install jspm@beta`). Please test it out if you can, feedback welcome.

**Comment by vanko-dev** [comment]
wow, works fine on teamcity. win server 12 and win 8.1. Thanks!

**Comment by martingust** [comment]
Works very good! This is helping us a lot and probably a lot of others. Big thanks @crisptrutski @brendonparker @guybedford great work!!!

**Comment by XVincentX** [comment]
That is great. When the beta will not be beta anymore?

**Comment by guybedford** [comment]
We're just waiting on Aurelia to be updated to work against the latest beta at the moment, being tracked in https://github.com/aurelia/bootstrapper/issues/10. If anyone is able to assist with that, it may well be very welcome there.

**Comment by amcdnl** [comment]
Confirmed this works with beta with my setup.

**Comment by pawelpawlow** [comment]
Confirmed, beta works. Can't wait for Aurelia to be updated.

**Comment by pawelpawlow** [comment]
@guybedford https://github.com/aurelia/bootstrapper/issues/10 was closed 20 days ago, but it looks like beta and aurelia still doesn't work together. Are there other issues that hold jspm beta to be not beta anymore?

**Comment by togakangaroo** [comment]
Was having the same issue so I `npm install jspm@beta --save`

Now

```
> npm run jspm:install

> MyApp@0.0.0 jspm:install W:\Surge\MyApp\MyApp
> jspm install

     Looking up npm:core-js
     Looking up npm:babel-core
     Looking up npm:classnames
     Looking up npm:formsy-react
     Looking up npm:lodash
     Looking up npm:babel-runtime
     Looking up npm:react
     Looking up npm:react-router
     Looking up npm:react-select-box
     Looking up npm:react-paginate
     Looking up npm:reflux
     Looking up npm:thee
     Updating registry cache...
     Looking up github:components/jquery

warn Error on lookup for npm:babel-runtime
     Invalid authentication details. Run jspm registry config npm to reconfigure.

err  Error looking up npm:babel-runtime.

warn Installation changes not saved.
```

**Comment by guybedford** [comment]
@togakangaroo can you share your `.npmrc` file structure? It would be a great help.

Also can you make sure you check the install `jspm@beta` to get the latest jspm-npm dependency within it (0.23.4)?

**Comment by togakangaroo** [comment]
Hi @guybedford my `~/.npmrc` contains only an _auth and email lines. I don't have an `.npmrc` in this project and as far as I know no other ones on my machine (except the default)

checking `node_modules/jspm/node_modules/jspm-npm/package.json` I show version 0.23.4.

looking at the `node_modules/jspm/package.json` I see that it *is* indeed the beta one

      "_resolved": "https://registry.npmjs.org/jspm/-/jspm-0.16.0-beta.3.tgz"

Anything else to check?

**Comment by togakangaroo** [comment]
For reference, created a blank directory with just the following

**Comment by togakangaroo** [comment]
@guybedford did that help? Still hoping to do something that will enable us to deploy

**Comment by guybedford** [comment]
@togakangaroo have you got authentication configured in your `~/.jspm/config` file for the npm registry?

**Comment by togakangaroo** [comment]
```
W:\> cat C:\Users\George\.npmrc
_auth=dG....Rs
email=g...r@gmail.com
```

**Comment by guybedford** [comment]
@togakangaroo this could be a character encoding issue with your npm credentials. To debug this, try adding a `console.log(this.auth)` statement to the line just before https://github.com/jspm/npm/blob/master/lib/npm.js#L57 in `node_modules/jspm-npm/lib/npm.js` within the jspm folder. That should then be fairly obvious how your username or password is being interpreted incorrectly. If it's a simple character encoding issue please let me know so I can correct it. We did just fix one of these in jspm-npm 0.23.4 as you said you're using, but do check again you're definitely running this version.

**Comment by togakangaroo** [comment]
```
> cat .\node_modules\jspm\node_modules\jspm-npm\package.json
{
  "name": "jspm-npm",
  "version": "0.23.4",
  ...
```
@guybedford not sure what it should be but after adding that line
```
> jspm install
{ username: 'togakangaroo', password: 't7....tl' }
     Looking up npm:babel-core
     Looking up npm:babel-runtime
     Looking up npm:core-js
     Looking up npm:formsy-react
     Looking up npm:classnames
     Looking up npm:jquery
     Looking up npm:moment
     Looking up npm:react-paginate
     Looking up npm:react
     Looking up npm:lodash
     Looking up npm:react-router
     Looking up npm:reflux
     Looking up npm:react-select-box
     Looking up npm:thee
     Updating registry cache...

warn Error on lookup for npm:react-router
     Invalid authentication details. Run jspm registry config npm to reconfigure.

err  Error looking up npm:react-router.

warn Installation changes not saved.
```

that certainly doesn't seem like a character encoding issue to me?

**Comment by guybedford** [comment]
Thanks for following up, so it is your correct password to npmjs.org there?

**Comment by togakangaroo** [comment]
@guybedford yes, it is.

**Comment by guybedford** [comment]
Are you completely sure? Did you try directly copying the password as written from that log statement, running `npm login` and copying the exact same text in to there? I'd be very surprised if that worked, as that's all we do in jspm anyway effectively.

**Comment by togakangaroo** [comment]
@guybedford I have now. Did this both from conemu/powershell and a basic cmd terminal and was able to log into the site. The `npm login` command also worked. Surprise!

Really wish I had time to put a day aside digging into this but I don't for at least a week or two.

**Comment by guybedford** [comment]
@togakangaroo sure if you do have time to look into it note that the only step running between jspm having that exact auth that you checked and the actual loopup is simply a GET request to `https://registry.npmjs.org/react-router` etc with those exact auth credentials set. So if you do have time to test it I would suggest attempting a curl along the lines of `curl -u togakangaroo:t7....tl https://registry.npmjs.org/react-router`. If that works then there's really nothing else to check after that provided the credentials that were logged previously are correct.

**Comment by togakangaroo** [comment]
For the record, I was able to get jspm@beta working (`0.16.0-beta.3`) but am *still* having trouble on our build server. It seems to get a lot further but still times out as before with errors such as

```
build	06-Aug-2015 13:34:14	warn Error on lookup for github:jspm/nodelibs-assert, retrying (1).
build	06-Aug-2015 13:34:14	     Error: fatal: failed to stat '.': No such file or directory
```

**Comment by guybedford** [comment]
@togakangaroo if the build server has persistent state this may be a cache issue, perhaps try running `jspm cc` to clear the cache there.
